### PR TITLE
[branch-4.2] fix(coredump thread): Coredump thread should not retry forever on exception

### PR DESCRIPTION
**Pull request by juliayakovlev**
If exceptions will happen 10 times during getting backtraces - fail the thread and don't run it forever

[Task](https://trello.com/c/Jbpm0ywF/2390-branch-20201-42-coredump-thread-should-not-retry-forever-on-unknown-exception)

## PR pre-checks (self review)

- [x] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [x] I didn't leave commented-out/debugging code
- [x] I added the relevant `backport` labels
- [ ] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [x] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [x] All new and existing unit tests passed (CI)
- [ ] I have updated the Readme/doc folder accordingly (if needed)

### CODE REVIEW

**Comment by fruch** [comment]
@juliayakovlev what's happen if we don't do this ? 

I was under the assumption we want exactly the opposite, to never kill this thread.

**Comment by juliayakovlev** [comment]
> @juliayakovlev what's happen if we don't do this ?
> 
> I was under the assumption we want exactly the opposite, to never kill this thread.

In case the coredump thread is stuck with exceptions, we don't need to keep it.
It was decided to try 10 times before fail the thread

File: unit_tests/test_cluster.py [hunk 1]
```diff
@@ -126,6 +132,145 @@ def test_search_system_suppressed_messages(self):  # pylint: disable=invalid-nam
         assert event_a["line_number"] == 6, 'Not expected event line number {}'.format(event_a["line_number"])
 
 
+def setup_test(func):
```
Comment (bentsi) [inline unit_tests/test_cluster.py:135]: This is not needed. Unit test framework does this for you.

**Review by bentsi — CHANGES_REQUESTED** [review]
use setUp and tearDown methods.

File: unit_tests/test_cluster.py (hunk 1 above)
Comment (fruch) [reply unit_tests/test_cluster.py:135]: (actually it's pytest that doesn't it nicely in our case)

Comment (juliayakovlev) [reply unit_tests/test_cluster.py]: Fixed

**Comment by bentsi** [comment]
```
19:21:27  >       assert "sdcm.remote.base.SSHConnectTimeoutError" == traceback, \
19:21:27                 f"Expected exception 'sdcm.remote.base.SSHConnectTimeoutError', got {traceback}"
19:21:27  E       AssertionError: Expected exception 'sdcm.remote.base.SSHConnectTimeoutError', got None
19:21:27  E       assert 'sdcm.remote....tTimeoutError' == None
19:21:27  E         -'sdcm.remote.base.SSHConnectTimeoutError'
19:21:27  E         +None
19:21:27  
19:21:27  unit_tests/test_decode_backtrace.py:216: AssertionError
```

**Comment by juliayakovlev** [comment]
@bentsi 
All unit tests are passed

File: unit_tests/dummy_remote.py [hunk 2]
```diff
@@ -41,6 +47,58 @@ def receive_files(src, dst):
...
+
+        if 'sudo coredumpctl --no-pager' in args[0]:
+            dummy_output = "No coredumps found"
+        else:
+            dummy_output = args[0]
+        return DummyOutput(dummy_output)
+
+    def raise_exception(self):
+        if self.test_exception:
+            while self.exceptions_amount:
```
Comment (enaydanov) [inline unit_tests/dummy_remote.py:72]: It's effectively `if`

File: unit_tests/test_decode_backtrace.py [hunk 3]
```diff
@@ -121,21 +135,149 @@ def test_04_decode_backtraces_core(self):
...
+        traceback = ""
+        if events:
+            last_event = events[-1]
+            if "traceback" in last_event:
+                print(f"Last event: {last_event['traceback']}")
+                traceback = last_event["traceback"].splitlines()[-1]
+        return traceback
+
+    def wait_for_all_exceptions_raised(self, exceptions_amount):
+        while self.node.remoter.run_number < exceptions_amount:
```
Comment (enaydanov) [inline unit_tests/test_decode_backtrace.py:205]: Same. It's not a loop. And why we compare `run_number` with `exceptions_amount`?

File: unit_tests/test_decode_backtrace.py [hunk 4]
```diff
@@ -121,21 +135,149 @@ def test_04_decode_backtraces_core(self):
...
+    def test_backtrace_thread_with_no_last_exception(self):
+        """
+        Test case when get backtraces function failed after an exception was raised 2 times and then continue
+        without failure.
+        get backtraces function should fail after an exception was raised 3 times. Because of just 2 exceptions will be
+        raised, expected that coredump thread continue to run.
+
+        """
+        exceptions_amount = 2
+        self.node.max_coredump_thread_exceptions = exceptions_amount+1
```
Comment (enaydanov) [inline unit_tests/test_decode_backtrace.py:270]: Add spaces around `+`

File: unit_tests/test_decode_backtrace.py [hunk 5]
```diff
@@ -121,21 +135,149 @@ def test_04_decode_backtraces_core(self):
...
+        self.node.remoter.test_exception = ValueError
+        self.node.remoter.exceptions_amount = exceptions_amount
+
+        self.node.start_backtrace_thread()
+
+        wait.wait_for(self.wait_for_all_exceptions_raised, timeout=5, exceptions_amount=exceptions_amount)
```
Comment (enaydanov) [inline unit_tests/test_decode_backtrace.py:276]: So, it used to organize a cycle with sleep inside?

Like:

```python
for _ in range(exceptions_amount):
    if self.node.remoter.run_number < min(exceptions_amount, 5):
        sleep(1)
```
???

**Review by enaydanov — CHANGES_REQUESTED** [review]

File: unit_tests/dummy_remote.py (hunk 2 above)
Comment (juliayakovlev) [reply unit_tests/dummy_remote.py:72]: The function raise_exception is not called "exceptions_amount" times. It raises needed amount of exceptions and exit. So it should be loop

Comment (enaydanov) [reply unit_tests/dummy_remote.py:72]: You can't raise exceptions in a loop :) The first one will break it.

File: sdcm/cluster.py [hunk 6]
```diff
@@ -441,6 +441,46 @@ def __init__(self, name, parent_cluster, ssh_login_info=None, base_logdir=None,
         self._kernel_version = None
         self._cassandra_stress_version = None
 
+        # Next 3 parameters are used and allowed to be changed by unit tests in unit_tests/test_decode_backtrace.py.
+        # Want to change it to reduce the test time
+        self._backtrace_lookup_period: int = 15
+        self._get_core_pids_func_retrying: int = 10
+        self._get_core_pids_func_sleep_timeout: int = 20
+        # If exceptions will happen MAX_COREDUMP_THREAD_EXCEPTIONS times - fail the thread and don't run it forever
```
[the remark this answers is not in the document]
Comment (juliayakovlev) [reply sdcm/cluster.py]: fixed

File: sdcm/cluster.py [hunk 7]
```diff
@@ -441,6 +441,46 @@ def __init__(self, name, parent_cluster, ssh_login_info=None, base_logdir=None,
...
+        self._get_core_pids_func_sleep_timeout = value
+
+    @property
+    def get_core_pids_func_retrying(self):
+        return self._get_core_pids_func_retrying
+
+    @get_core_pids_func_retrying.setter
+    def get_core_pids_func_retrying(self, value):
+        self._get_core_pids_func_retrying = value
+
```
[the remark this answers is not in the document]
Comment (juliayakovlev) [reply sdcm/cluster.py]: fixed

File: sdcm/cluster.py [hunk 8]
```diff
@@ -985,59 +1025,73 @@ def get_backtraces(self):
...
+            self.termination_event.wait(self.backtrace_lookup_period)
+            try:
+                last_exc = self.get_backtraces()
+                exceptions_count = 0 if last_exc == None else exceptions_count + 1
+            except Exception as exc:
+                exceptions_count += 1
+                last_exc = exc
+                self.log.error(f"Following error occurred during getting back traces: {str(last_exc)}")
+
+            # If exceptions will happen MAX_COREDUMP_THREAD_EXCEPTIONS times - fail the thread and don't run it forever
```
[the remark this answers is not in the document]
Comment (juliayakovlev) [reply sdcm/cluster.py]: removed

File: unit_tests/dummy_remote.py [hunk 9]
```diff
@@ -41,6 +47,58 @@ def receive_files(src, dst):
         return True
 
 
+class DummyRemoteWithException(DummyRemote):
+
+    def __init__(self) -> None:
+        # Hold exception object that should be raised
+        self._test_exception: Exception or None = None
```
[the remark this answers is not in the document]
Comment (juliayakovlev) [reply unit_tests/dummy_remote.py]: changed

File: unit_tests/dummy_remote.py [hunk 10]
```diff
@@ -41,6 +47,58 @@ def receive_files(src, dst):
...
+    def exceptions_amount(self, amount):
+        self._exception_amount = amount
+
+    @property
+    def run_number(self):
+        return self._run_number
+
+    @run_number.setter
+    def run_number(self, amount):
+        self._run_number = amount
```
[the remark this answers is not in the document]
Comment (juliayakovlev) [reply unit_tests/dummy_remote.py]: removed

File: unit_tests/test_decode_backtrace.py [hunk 11]
```diff
@@ -121,21 +135,149 @@ def test_04_decode_backtraces_core(self):
...
         Setup.BACKTRACE_DECODING = True
+        self.open_event_file()
 
         self.monitor_node.start_decode_on_monitor_node_thread()
         self.node.system_log = os.path.join(os.path.dirname(__file__), 'test_data', 'system_core.log')
 
-        self.node._read_system_log_and_publish_events()
-
+        self.node._read_system_log_and_publish_events(start_from_beginning=True)
+        # Wait for events been published
```
[the remark this answers is not in the document]
Comment (juliayakovlev) [reply unit_tests/test_decode_backtrace.py:147]: fixed

File: sdcm/cluster.py [hunk 12]
```diff
@@ -985,59 +1025,73 @@ def get_backtraces(self):
...
         """
-
+        exceptions_count = 0
         while not self.termination_event.isSet():
-            self.termination_event.wait(15)
-            self.get_backtraces()
+            self.termination_event.wait(self.backtrace_lookup_period)
+            try:
+                last_exc = self.get_backtraces()
+                exceptions_count = 0 if last_exc == None else exceptions_count + 1
```
[the remark this answers is not in the document]
Comment (juliayakovlev) [reply sdcm/cluster.py]: fixed

File: sdcm/cluster.py [hunk 13]
```diff
@@ -985,59 +1025,73 @@ def get_backtraces(self):
...
+            except Exception as exc:
+                exceptions_count += 1
+                last_exc = exc
+                self.log.error(f"Following error occurred during getting back traces: {str(last_exc)}")
+
+            # If exceptions will happen MAX_COREDUMP_THREAD_EXCEPTIONS times - fail the thread and don't run it forever
+            if exceptions_count == self.max_coredump_thread_exceptions:
+                raise last_exc
```
[the remark this answers is not in the document]
Comment (juliayakovlev) [reply sdcm/cluster.py]: fixed

File: unit_tests/dummy_remote.py [hunk 14]
```diff
@@ -23,9 +23,15 @@
...
         self.stdout = stdout
+        self.exit_status = 0
+        self.stderr = ""
 
 
 class DummyRemote:
+
+    def __init__(self) -> None:
+        pass
+
```
[the remark this answers is not in the document]
Comment (juliayakovlev) [reply unit_tests/dummy_remote.py]: removed

File: unit_tests/dummy_remote.py (hunk 2 above)
Comment (juliayakovlev) [reply unit_tests/dummy_remote.py]: you are right. fixed

File: unit_tests/test_decode_backtrace.py (hunk 3 above)
Comment (juliayakovlev) [reply unit_tests/test_decode_backtrace.py]: changed

File: unit_tests/test_decode_backtrace.py (hunk 4 above)
Comment (juliayakovlev) [reply unit_tests/test_decode_backtrace.py]: fixed

File: unit_tests/test_decode_backtrace.py (hunk 5 above)
Comment (juliayakovlev) [reply unit_tests/test_decode_backtrace.py]: changed

**Review by enaydanov — CHANGES_REQUESTED** [review]

File: sdcm/cluster.py [hunk 15]
```diff
@@ -985,59 +992,76 @@ def get_backtraces(self):
...
+            try:
+                last_exc = self.get_backtraces()
+            except Exception as exc:
+                last_exc = exc
+
+            if last_exc is None:
+                exceptions_count = 0
+                continue
+
+            self.log.error(f"Following error occurred during getting back traces: %s", last_exc)
```
[the remark this answers is not in the document]
Comment (juliayakovlev) [reply sdcm/cluster.py]: changed

File: sdcm/cluster.py [hunk 16]
```diff
@@ -985,59 +992,76 @@ def get_backtraces(self):
...
+        return _get_core_pids_inner()
 
     @raise_event_on_failure
     def backtrace_thread(self):
         """
         Keep reporting new coredumps found, every 30 seconds.
         """
-
+        exceptions_count = 0
         while not self.termination_event.isSet():
```
[the remark this answers is not in the document]
Comment (juliayakovlev) [reply sdcm/cluster.py]: changed

File: sdcm/cluster.py [hunk 17]
```diff
@@ -985,59 +992,76 @@ def get_backtraces(self):
...
                 if core_pid in self._discovered_backtraces:
                     continue
                 if len(self._discovered_backtraces) > self._maximum_number_of_cores_to_publish:
                     break
                 self._notify_backtrace(core_pid)
                 self._discovered_backtraces.append(core_pid)
+            return None
         except NETWORK_EXCEPTIONS as exc:
             self.log.error(f"Following network error occurred during getting back traces: {str(exc)}")
+            return exc
```
[the remark this answers is not in the document]
Comment (juliayakovlev) [reply sdcm/cluster.py:1009]: done

File: sdcm/cluster.py [hunk 18]
```diff
@@ -441,6 +441,13 @@ def __init__(self, name, parent_cluster, ssh_login_info=None, base_logdir=None,
         self._kernel_version = None
         self._cassandra_stress_version = None
 
+        # Next 3 parameters are used and allowed to be changed by unit tests in unit_tests/test_decode_backtrace.py.
+        # Want to change it to reduce the test time
+        self.backtrace_lookup_period: int = 15
+        self.get_core_pids_func_retrying: int = 10
+        self.get_core_pids_func_sleep_timeout: int = 20
+        self.max_coredump_thread_exceptions: int = 10
```
[the remark this answers is not in the document]
Comment (juliayakovlev) [reply sdcm/cluster.py]: fxed

File: sdcm/cluster.py [hunk 19]
```diff
@@ -441,6 +442,12 @@ def __init__(self, name, parent_cluster, ssh_login_info=None, base_logdir=None,
         self._kernel_version = None
         self._cassandra_stress_version = None
 
+        # Next 3 parameters are used and allowed to be changed by unit tests in unit_tests/test_decode_backtrace.py.
+        # Want to change it to reduce the test time
+        self.backtrace_lookup_period: int = 15
+        self.get_core_pids_func_retrying: int = 10
+        self.get_core_pids_func_sleep_timeout: int = 20
+
```
Comment (enaydanov) [inline sdcm/cluster.py:450]: I meant these variables too. Please, move them to class level.

File: sdcm/cluster.py [hunk 20]
```diff
@@ -350,6 +350,7 @@ class BaseNode(AutoSshContainerMixin, WebDriverContainerMixin):  # pylint: disab
     CQL_PORT = 9042
     MANAGER_AGENT_PORT = 10001
     MANAGER_SERVER_PORT = 56080
+    max_coredump_thread_exceptions: int = 10
```
Comment (enaydanov) [inline sdcm/cluster.py:353]: Please, add one empty line before to separate this block.

**Review by enaydanov — CHANGES_REQUESTED** [review]

File: sdcm/cluster.py (hunk 20 above)
Comment (juliayakovlev) [reply sdcm/cluster.py:354]: done

File: sdcm/cluster.py (hunk 19 above)
Comment (juliayakovlev) [reply sdcm/cluster.py]: done

**Review by enaydanov — APPROVED** [review]
Looks good.

**Pull request merged**